app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java

```diff
@@ -331,6 +358,121 @@ public void onComplete() { }
+                builder.create().show();
+                break;
+            default:
+                return super.onOptionsItemSelected(item);
+        }
+        return true;
+    }
+
+    public void removeWatchedStreams(final boolean removePartiallyWatched) {
+        if (removeWatchedDisposable != null && !removeWatchedDisposable.isDisposed()) {
+            // already running
+            return;
+        }
+        showLoading();
+
+        removeWatchedDisposable = Flowable.just(Flowable.just(removePartiallyWatched,
+                playlistManager.getPlaylistStreams(playlistId).blockingFirst()))
+                .subscribeOn(Schedulers.io())
+                .map(flow -> {
+                    boolean localRemovePartiallyWatched = (boolean) flow.blockingFirst();
+                    List<PlaylistStreamEntry> playlist
+                            = (List<PlaylistStreamEntry>) flow.blockingLast();
+                    HistoryRecordManager recordManager = new HistoryRecordManager(getContext());
+                    Iterator<PlaylistStreamEntry> playlistIter = playlist.iterator();
+                    Iterator<StreamHistoryEntry> historyIter = recordManager
+                            .getStreamHistorySortedById().blockingFirst().iterator();
+                    List<PlaylistStreamEntry> notWatchedItems = new ArrayList<>();
+                    Iterator<StreamStateEntity> streamStatesIter = null;
+                    boolean thumbnailVideoRemoved = false;
+
+                    if (!localRemovePartiallyWatched) {
+                        streamStatesIter = recordManager.loadLocalStreamStateBatch(playlist)
+                                .blockingGet().iterator();
+                    }
+
+                    // already sorted by ^ getStreamHistorySortedById(), binary search can be used
+                    ArrayList<Long> historyStreamIds = new ArrayList<>();
+                    while (historyIter.hasNext()) {
+                        historyStreamIds.add(historyIter.next().getStreamId());
+                    }
+
+                    if (localRemovePartiallyWatched) {
+                        while (playlistIter.hasNext()) {
+                            PlaylistStreamEntry playlistItem = playlistIter.next();
+                            int indexInHistory = Collections.binarySearch(historyStreamIds,
+                                    playlistItem.getStreamId());
+
+                            if (indexInHistory < 0) {
+                                notWatchedItems.add(playlistItem);
+                            } else if (!thumbnailVideoRemoved
+                                    && playlistManager.getPlaylistThumbnail(playlistId)
+                                    .equals(playlistItem.getStreamEntity().getThumbnailUrl())) {
+                                thumbnailVideoRemoved = true;
+                            }
+                        }
+                    } else {
+                        boolean hasState = false;
```

Comment:
Declare variables right where you first use them. Redeclaring a variable every cycle of the while loop does not slow down anything, since declaring is just telling the compiler "I need to have this memory at this point", so memory is really allocated only once.

Reply:
Fixed in db335d5